### PR TITLE
cop: move the logic of limiting heavy tasks out of runner

### PR DESCRIPTION
**Pull request by sticnarf**
###  What have you changed?

@andylokandy suggets not to put the logic of limiting heavy tasks in `BatchExecutorRunner` because the runner should actually not care about this extra scheduling strategy.

This PR implements a util Future that limits the time used by a Future without getting a semaphore permit. Then, the logic only occurs in the coprocessor endpoint and make the code much clearer.

###  What is the type of the changes?

- Engineering (engineering change which doesn't change any feature or fix any issue)

###  How is the PR tested?

- Unit test

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?

No

###  Does this PR affect `tidb-ansible`?

No

### CODE REVIEW

File: components/tidb_query/src/batch/runner.rs [hunk 1]
```diff
@@ -40,12 +39,6 @@ pub struct BatchExecutorsRunner<SS> {
     // TODO: Deprecate it using a better deadline mechanism.
     deadline: Deadline,
```
Comment (andylokandy) [inline components/tidb_query/src/batch/runner.rs]: Can deadline also be extracted?

File: src/coprocessor/metrics.rs [hunk 2]
```diff
@@ -67,6 +68,30 @@ lazy_static! {
...
+    pub static ref COPR_WAITING_FOR_SEMAPHORE: IntGauge = register_int_gauge!(
+        "tikv_coprocessor_waiting_for_semaphore",
+        "The number of tasks waiting for the semaphore"
+    )
+    .unwrap();
+}
+
+make_static_metric! {
+    pub label_enum AcquireSemaphoreType {
+        unacquired,
```
Comment (andylokandy) [inline src/coprocessor/metrics.rs]: Why `acquiring` is removed?

File: src/coprocessor/time_limiter.rs [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+use std::future::Future;
+use std::marker::PhantomData;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use std::time::{Duration, Instant};
+use tokio::sync::{Semaphore, SemaphorePermit};
+
+use super::metrics::*;
+
+/// Limits the time used by `fut` before a semaphore permit is acquired.
```
Comment (andylokandy) [inline src/coprocessor/time_limiter.rs]:
```suggestion
/// Limits the time spent on executing `fut` before forcing to acquire a semaphore permit.
```

File: src/coprocessor/time_limiter.rs [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+///
+/// The future `fut` can always run for at least `time_limit_without_permit`,
+/// but it needs to acquire a permit from the semaphore before it can continue.
+pub fn limit_time<'a, F: Future + 'a>(
```
Comment (andylokandy) [inline src/coprocessor/time_limiter.rs]: I think the name `limit_time` is vague. It is not a mechanism to limit the execution time but a concurrent limiter (time limit is its method, but not the purpose). Maybe we can call it `ConcurrentGate` or `ConcurrentLimiter`.

File: src/coprocessor/time_limiter.rs [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+{
+    #[pin]
+    permit_fut: PF,
+    #[pin]
+    fut: F,
+    time_limit_without_permit: Duration,
+    execution_time: Duration,
+    permit: Option<SemaphorePermit<'a>>,
+    // Whether the task is waiting for a semaphore. Used for tracking metrics only.
+    waiting: bool,
```
Comment (andylokandy) [inline src/coprocessor/time_limiter.rs]: We can make them into a state enum: Limitation: NotLimited, Acquiring, Acuqired(SemaphorePermit)

File: src/coprocessor/metrics.rs (hunk 2 above)
Comment (sticnarf) [reply src/coprocessor/metrics.rs]: Actually there isn't `acquring` before. `COPR_WAITING_FOR_SEMAPHORE` refers to `acquiring`.

File: components/tidb_query/src/batch/runner.rs (hunk 1 above)
Comment (sticnarf) [reply components/tidb_query/src/batch/runner.rs]: We can. But I'm hesitating integrating to `ConcurrencyLimiter` or adding another interceptor for it.

This PR is fairly large now. I don't intend to add more to it.

**Comment by sticnarf** [comment]
@andylokandy PTAL again

**Pull request merged**